src/dense.jl

```diff
@@ -0,0 +1,243 @@
+"""
+    AutoChainRules{RC}
+
+Chooses any AD library based on [ChainRulesCore.jl](https://github.com/JuliaDiff/ChainRulesCore.jl) (see the list [here](https://juliadiff.org/ChainRulesCore.jl/stable/index.html#ChainRules-roll-out-status)).
+
+# Fields
+
+- `ruleconfig::RC`: a [`ChainRulesCore.RuleConfig`](https://juliadiff.org/ChainRulesCore.jl/stable/rule_author/superpowers/ruleconfig.html) object.
+
+# Constructor
+
+    AutoChainRules(; ruleconfig)
+"""
+Base.@kwdef struct AutoChainRules{RC} <: AbstractADType
+    ruleconfig::RC
+end
+
+mode(::AutoChainRules) = ForwardOrReverseMode()  # specialized in the extension
+
+"""
+    AutoDiffractor
+
+Chooses [Diffractor.jl](https://github.com/JuliaDiff/Diffractor.jl).
+
+# Constructor
+
+    AutoDiffractor()
+"""
+struct AutoDiffractor <: AbstractADType end
+
+mode(::AutoDiffractor) = ForwardOrReverseMode()
+
+"""
+    AutoEnzyme{M}
+
+Chooses [Enzyme.jl](https://github.com/EnzymeAD/Enzyme.jl).
+
+# Fields
+
+- `mode::M`: can be either 
+  - an object subtyping `EnzymeCore.Mode` (like `EnzymeCore.Forward` or `EnzymeCore.Reverse`) if a specific mode is required
+  - `nothing` to choose the best mode automatically
+
+# Constructors
+
+    AutoEnzyme(; mode=nothing)
+"""
+Base.@kwdef struct AutoEnzyme{M} <: AbstractADType
+    mode::M = nothing
+end
+
+mode(::AutoEnzyme) = ForwardOrReverseMode()  # specialized in the extension
+
+"""
+    AutoFastDifferentiation
+
+Chooses [FastDifferentiation.jl](https://github.com/brianguenter/FastDifferentiation.jl).
+
+# Constructor
+
+    AutoFastDifferentiation()
+"""
+struct AutoFastDifferentiation <: AbstractADType end
+
+mode(::AutoFastDifferentiation) = SymbolicMode()
+
+"""
+    AutoFiniteDiff{T1,T2,T3}
+
+Chooses [FiniteDiff.jl](https://github.com/JuliaDiff/FiniteDiff.jl).
+
+# Fields
+
+- `fdtype::T1`: finite difference type
+- `fdjtype::T2`: finite difference type for the Jacobian
+- `fdhtype::T3`: finite difference type for the Hessian
+
+# Constructor
+
+    AutoFiniteDiff(; fdtype=Val(:forward), fdjtype=fdtype, fdhtype=Val(:hcentral))
+"""
+Base.@kwdef struct AutoFiniteDiff{T1, T2, T3} <: AbstractADType
+    fdtype::T1 = Val(:forward)
+    fdjtype::T2 = fdtype
+    fdhtype::T3 = Val(:hcentral)
+end
+
+mode(::AutoFiniteDiff) = FiniteDifferencesMode()
+
+"""
+    AutoFiniteDifferences{T}
+
+Chooses [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
+
+# Fields
+
+- `fdm::T`: a [`FiniteDifferenceMethod`](https://juliadiff.org/FiniteDifferences.jl/stable/pages/api/#FiniteDifferences.FiniteDifferenceMethod), constructed for instance with `FiniteDifferences.central_fdm`. 
+
+# Constructor
+
+    AutoFiniteDifferences(; fdm)
+"""
+Base.@kwdef struct AutoFiniteDifferences{T} <: AbstractADType
+    fdm::T
```

Comment:
I think you might have missed to add a default here?

Reply:
See the comment at the top, I voluntarily removed it because picking "nothing" doesn't make much sense here. Like for AutoChainRules, I feel like the field of AutoFiniteDifferences has to be a specific type anyway